fds/run.py

```diff
@@ -1,78 +1,169 @@
+import enum
 from shutil import which
 import PyInquirer
 import requests
+from pathlib import Path
 import sys
 
 from fds.domain.commands import Commands
 from fds.logger import Logger
 from fds.services.dvc_service import DVCService
 from fds.services.fds_service import FdsService
+from fds.services.types import InnerService
 from fds.services.git_service import GitService
 from fds.services.pretty_print import PrettyPrint
-from fds.utils import execute_command
+from fds.utils import execute_command, rerun_in_new_shell_and_exit
 from .__init__ import __version__
 
 
-class Run(object):
-    def __init__(self, arguments: dict):
-        self.logger = Logger.get_logger("fds.Run")
-        self.arguments = arguments
-        self.service = FdsService(GitService(), DVCService())
-        self.printer = PrettyPrint()
+class HooksRunner(object):
+    class ExitCodes(enum.IntFlag):
+        OK = 0
+        DVC_INSTALL_FAILED = 2**0
+        GIT_INSTALL_FAILED = 2**1
+        FDS_UPDATE_FAILED = 2**1
+        GIT_INITIALIZE_FAILED = 2**3
+        DVC_INITIALIZE_FAILED = 2**4
 
-    def pre_execute_hook(self):
-        # Check if dvc is installed
-        ret_code = 0
-        if which("dvc") is None:
-            ret_code = 1
-            self.printer.error("dvc executable is not installed or found")
-            questions = [
-                {
-                    'type': 'confirm',
-                    'message': 'Should we install dvc [https://dvc.org/] for you right now?\n' +
-                               '  Will install using `pip install dvc<3`',
-                    'name': 'install',
-                    'default': False,
-                },
-            ]
-            answers = PyInquirer.prompt(questions)
-            if answers["install"]:
-                execute_command(["pip install 'dvc<3'"], shell=True, capture_output=False)
-                ret_code = 0
-            else:
-                # Provide instructions
-                self.printer.warn("You can install dvc manually from https://dvc.org/doc/install")
+    def __init__(
+        self,
+        service: FdsService,
+        printer: PrettyPrint,
+        logger: Logger,
+    ):
+        self.service = service
+        self.printer = printer
+        self.logger = logger
+
+        self.pre_execute_hooks = [
+            (self.ExitCodes.DVC_INSTALL_FAILED.value, self._ensure_dvc_installed),
+            (self.ExitCodes.GIT_INSTALL_FAILED.value, self._ensure_git_installed),
+            (self.ExitCodes.FDS_UPDATE_FAILED.value, self._ensure_fds_updated),
+            (self.ExitCodes.GIT_INITIALIZE_FAILED.value, self._ensure_git_initialized),
+            (self.ExitCodes.DVC_INITIALIZE_FAILED.value, self._ensure_dvc_initialized),
+        ]
```

Comment:
Nice approach!

Reply:
I think that improves maintainability and allows for expansion (eg more pre_execute_hooks, or add post_execute_hooks, etc)